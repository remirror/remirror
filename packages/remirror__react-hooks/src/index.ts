export * from './react-hook-utils';
export * from './use-editor-event';
export * from './use-editor-focus';
export * from './use-emoji';
export * from './use-history';
export * from './use-hover';
export * from './use-keymap';
export * from './use-keymaps';
export * from './use-mention';
export * from './use-mention-atom';
export * from './use-menu-navigation';
export * from './use-multi-positioner';
export * from './use-positioner';
export * from './use-suggest';
