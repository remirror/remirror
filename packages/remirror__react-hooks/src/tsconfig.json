{
  "__AUTO_GENERATED__": [
    "To update the configuration edit the following field.",
    "`package.json > @remirror > tsconfigs > 'src'`",
    "",
    "Then run: `pnpm -w generate:ts`"
  ],
  "extends": "../../../support/tsconfig.base.json",
  "compilerOptions": {
    "types": [],
    "declaration": true,
    "composite": true,
    "noEmit": false,
    "emitDeclarationOnly": true,
    "outDir": "../dist-types"
  },
  "include": ["./"],
  "references": [
    {
      "path": "../../remirror__core/src"
    },
    {
      "path": "../../remirror__core-constants/src"
    },
    {
      "path": "../../remirror__core-helpers/src"
    },
    {
      "path": "../../remirror__extension-emoji/src"
    },
    {
      "path": "../../remirror__extension-events/src"
    },
    {
      "path": "../../remirror__extension-history/src"
    },
    {
      "path": "../../remirror__extension-mention/src"
    },
    {
      "path": "../../remirror__extension-mention-atom/src"
    },
    {
      "path": "../../remirror__extension-positioner/src"
    },
    {
      "path": "../../remirror__react-core/src"
    },
    {
      "path": "../../remirror__react-utils/src"
    },
    {
      "path": "../../multishift/src"
    },
    {
      "path": "../../remirror__pm/src"
    }
  ]
}
