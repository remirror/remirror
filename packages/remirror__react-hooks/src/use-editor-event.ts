import type { GetHandler, StringKey } from '@remirror/core';
import { EventsExtension, EventsOptions } from '@remirror/extension-events';
import { useExtensionEvent } from '@remirror/react-core';

/**
 * A hook for subscribing to events from the editor.
 */
export function useEditorEvent<Key extends StringKey<GetHandler<EventsOptions>>>(
  event: Key,
  handler: NonNullable<GetHandler<EventsOptions>[Key]>,
): void {
  useExtensionEvent(EventsExtension, event, handler);
}
