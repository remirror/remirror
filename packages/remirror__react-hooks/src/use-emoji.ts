import { useCallback, useMemo, useState } from 'react';
import { isEmptyArray, take } from '@remirror/core';
import {
  EmojiExtension,
  EmojiSuggestHandler,
  EmojiSuggestHandlerCommand,
  EmojiSuggestHandlerProps,
  FlatEmoji,
} from '@remirror/extension-emoji';
import { useExtensionEvent, useHelpers } from '@remirror/react-core';

import {
  MenuNavigationOptions,
  useMenuNavigation,
  UseMenuNavigationReturn,
} from './use-menu-navigation';

export interface FlatEmojiWithUrl extends FlatEmoji {
  /**
   * The svg url for CDN access.
   */
  url: string;
}

export interface EmojiState extends Pick<EmojiSuggestHandlerProps, 'range' | 'query' | 'exit'> {
  /**
   * The list of emoji generated by the query.
   *
   * @defaultValue []
   */
  list: FlatEmojiWithUrl[];

  /**
   * The command to run to replace the query with the request emoji.
   *
   * @defaultValue undefined
   */
  apply: EmojiSuggestHandlerCommand;
}

export interface UseEmojiProps extends MenuNavigationOptions {}

export interface UseEmojiReturn extends UseMenuNavigationReturn<FlatEmojiWithUrl> {
  /**
   * The state of the current query, only available when active.
   */
  state: EmojiState | null;
}

const emptyEmoji: never[] = [];

/**
 * This hook provides the state for setting up an emoji state change handler. It
 * applies the keybindings and the required change handlers.
 */
export function useEmoji(props: UseEmojiProps = {}): UseEmojiReturn {
  const { direction, dismissKeys, focusOnClick, submitKeys } = props;
  const [state, setState] = useState<EmojiState | null>(null);
  const helpers = useHelpers();
  const items = state?.list ?? emptyEmoji;
  const isOpen = !!state;

  const onDismiss = useCallback(() => {
    if (!state) {
      return false;
    }

    // Ignore the current mention so that it doesn't show again for this
    // matching area
    helpers
      .getSuggestMethods()
      .addIgnored({ from: state.range.from, name: 'emoji', specific: true });

    setState({ ...state, exit: true });
    return true;
  }, [helpers, state]);

  const onSubmit = useCallback(
    (emoji: FlatEmojiWithUrl) => {
      if (!state || isEmptyArray(state.list)) {
        return false;
      }

      state.apply(emoji.emoji);

      return true;
    },
    [state],
  );

  const menu = useMenuNavigation<FlatEmojiWithUrl>({
    items,
    isOpen,
    onDismiss,
    onSubmit,
    direction,
    dismissKeys,
    focusOnClick,
    submitKeys,
  });
  const { setIndex } = menu;

  const onChange: EmojiSuggestHandler = useCallback(
    (props) => {
      const { change, exit, query, moji, apply, range } = props;

      if (change) {
        setIndex(0);
        setState({
          list: take(moji.search(query), 20).map((emoji) => ({ ...emoji, url: moji.url(emoji) })),
          apply: (code) => {
            setState(null);
            return apply(code);
          },
          range,
          query,
          exit,
        });
      }

      if (exit) {
        setState(null);
      }
    },
    [setIndex],
  );

  // Add the change handler to the emoji state.
  useExtensionEvent(EmojiExtension, 'suggestEmoji', onChange);

  return useMemo(() => ({ ...menu, state }), [menu, state]);
}
