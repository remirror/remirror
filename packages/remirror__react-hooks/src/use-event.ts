import type { GetHandler, StringKey } from '@remirror/core';
import { EventsExtension, EventsOptions } from '@remirror/extension-events';
import { useExtensionEvent } from '@remirror/react-core';

/**
 * A hook for subscribing to events from the editor.
 */
export function useEvent<Key extends StringKey<GetHandler<EventsOptions>>>(
  event: Key,
  handler: NonNullable<GetHandler<EventsOptions>[Key]>,
): void {
  useExtensionEvent(EventsExtension, event, handler);
}

/**
 * @deprecated prefer `useEvent`
 */
export const useEvents = useEvent;
