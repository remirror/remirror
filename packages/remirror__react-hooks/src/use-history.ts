import type { GetHandler, StringKey } from '@remirror/core';
import { HistoryExtension, HistoryOptions } from '@remirror/extension-history';
import { useExtensionEvent } from '@remirror/react-core';

/**
 * A hook which is called every time an undo or redo event is triggered from
 * within the ProseMirror history extension.
 *
 * @remarks
 *
 * `handler` should be a memoized function.
 */
export function useHistory<Key extends StringKey<GetHandler<HistoryOptions>>>(
  event: Key,
  handler: NonNullable<GetHandler<HistoryOptions>[Key]>,
): void {
  useExtensionEvent(HistoryExtension, event, handler);
}
