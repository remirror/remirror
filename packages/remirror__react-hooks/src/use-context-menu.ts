import { ContextMenuEventHandler } from '@remirror/extension-events';

import { useEditorEvent } from './use-editor-event';

/**
 * A hook which listens to context menu events.
 *
 * In order to fully override the context menu events when they occur in the
 * editor make sure to call `event.preventDefault()` this will allow you to
 * replace the default context menu with your own version.
 */
export function useContextMenu(handler: ContextMenuEventHandler): void {
  useEditorEvent('contextmenu', handler);
}
