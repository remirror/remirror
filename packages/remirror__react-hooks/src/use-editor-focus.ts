import { useCallback, useState } from 'react';
import { FocusType, isElementDomNode } from '@remirror/core';
import { useRemirrorContext } from '@remirror/react-core';

import { useEditorEvent } from './use-editor-event';

export interface UseEditorFocusProps {
  /**
   * The elements that can be focused without setting `isFocused` to false.
   */
  ignoredElements?: Array<Element | null>;

  /**
   * Set this to true if you want to also update the focus value when the user
   * focuses on other windows or tabs (outside of the current DOM).
   *
   * @defaultValue false
   */
  blurOnInactive?: boolean;
}

/**
 * Keep track of the editor focus.
 *
 * Returns a focused value which is updated whenever the editor focus changes.
 *
 * When `true`, the editor is focused when `false` the editor is not focused.
 */
export function useEditorFocus(
  props: UseEditorFocusProps = {},
): [isFocused: boolean, focus: (position?: FocusType) => void] {
  const { ignoredElements = [], blurOnInactive = false } = props;

  // Get the view from the context and use it to calculate the initial focus.
  const { view, commands } = useRemirrorContext();

  // Create the initial state with the current view focus.
  const [isFocused, setIsFocused] = useState(() => view.hasFocus());

  // Listen to blur events and set focused to false in those instances.
  useEditorEvent(
    'blur',
    useCallback(
      (event: FocusEvent) => {
        const focusedElement = isElementDomNode(event.relatedTarget)
          ? event.relatedTarget
          : document.activeElement;
        const ignoreBlur = !blurOnInactive && !focusedElement;

        if (ignoreBlur || view.dom.contains(focusedElement)) {
          return false;
        }

        for (const element of ignoredElements) {
          if (element?.contains(focusedElement)) {
            return false;
          }
        }

        setIsFocused(false);

        return false;
      },
      [blurOnInactive, ignoredElements, view.dom],
    ),
  );

  // Listen to focus events and set focused to true in those instances.
  useEditorEvent(
    'focus',
    useCallback(
      (_: FocusEvent) => {
        if (isFocused) {
          return false;
        }

        setIsFocused(true);

        return false;
      },
      [isFocused],
    ),
  );

  return [isFocused, commands.focus];
}
