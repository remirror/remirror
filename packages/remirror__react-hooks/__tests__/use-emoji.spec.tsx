import { act, renderHook } from '@testing-library/react-hooks';
import { RemirrorTestChain } from 'jest-remirror';
import React, { FC } from 'react';
import { EmojiExtension } from 'remirror/extensions';
import data from 'svgmoji/emoji.json';
import { act as renderAct, strictRender } from 'testing/react';
import { createReactManager, Remirror, useRemirrorContext } from '@remirror/react-core';

import { useEmoji, UseEmojiReturn } from '../src/use-emoji';

function createChain() {
  const manager = createReactManager(() => [new EmojiExtension({ data })]);
  const chain = RemirrorTestChain.create(manager);
  const { doc, p } = chain.nodes;

  const Wrapper: FC<{ children?: React.ReactNode }> = ({ children }) => (
    <Remirror manager={manager} initialContent={[doc(p('Initial content ')), 'end']} autoRender>
      {children}
    </Remirror>
  );

  return { chain, Wrapper };
}

function acts(methods: Array<() => void | undefined>) {
  for (const method of methods) {
    act(method);
  }
}

describe('useEmoji', () => {
  it('should respond to emoji changes', () => {
    const { chain, Wrapper } = createChain();

    const { result } = renderHook(() => useEmoji(), {
      wrapper: Wrapper,
    });

    expect(result.current?.state).toBeNull();

    act(() => {
      chain.insertText(':a');
    });

    expect(result.current?.state).toEqual({
      range: { from: 17, to: 19, cursor: 19 },
      apply: expect.any(Function),
      exit: false,
      list: expect.any(Array),
      query: 'a',
    });

    expect(result.current?.index).toBe(0);

    expect(result.current?.state?.list.length).toBeGreaterThan(0);
  });

  it('should correctly add the emoji when the command is called', () => {
    const { Wrapper, chain } = createChain();

    const { result } = renderHook(() => useEmoji(), {
      wrapper: Wrapper,
    });

    acts([
      () => {
        chain.insertText(':red_heart');
      },
      () => result.current?.state?.apply(result.current.state?.list[0]?.emoji ?? ''),
    ]);

    expect(chain.innerHTML).toMatchSnapshot();
  });

  it('should clear suggestions when the `Escape` key is pressed', () => {
    const { chain, Wrapper } = createChain();

    const { result } = renderHook(() => useEmoji(), {
      wrapper: Wrapper,
    });

    acts([
      () => {
        chain.insertText(':');
      },
      () => {
        chain.press('Escape');
      },
    ]);

    expect(result.current?.state?.exit).toBe(true);

    acts([
      () => {
        chain.insertText('a');
      },
    ]);

    expect(result.current?.state).toBeNull();
  });

  it('should choose selection when `Enter` key is pressed', () => {
    const { Wrapper, chain } = createChain();

    renderHook(() => useEmoji(), {
      wrapper: Wrapper,
    });

    acts([
      () => {
        chain.insertText(':');
      },
      () => {
        chain.press('Enter');
      },
    ]);

    expect(chain.innerHTML).toMatchSnapshot();
  });

  it('should update index when `Arrow` keys are used', () => {
    const { Wrapper, chain } = createChain();

    const { result } = renderHook(() => useEmoji(), {
      wrapper: Wrapper,
    });

    acts([
      () => {
        chain.insertText(':');
      },
      () => {
        chain.press('ArrowDown');
      },
    ]);

    expect(result.current?.index).toBe(1);

    acts([
      () => {
        chain.press('ArrowUp');
      },
      () => {
        chain.press('ArrowUp');
      },
    ]);

    expect(result.current?.index).toBe(result.current!.state!.list.length - 1);
  });

  it('should update index when `Arrow` keys are used with `horizontal` direction', () => {
    const { Wrapper, chain } = createChain();

    const { result } = renderHook(() => useEmoji({ direction: 'horizontal' }), {
      wrapper: Wrapper,
    });

    acts([
      () => {
        chain.insertText(':');
      },
      () => {
        chain.press('ArrowRight');
      },
    ]);

    expect(result.current?.index).toBe(1);

    acts([
      () => {
        chain.press('ArrowLeft');
      },
      () => {
        chain.press('ArrowLeft');
      },
    ]);

    expect(result.current?.index).toBe(result.current!.state!.list.length - 1);
  });

  it('can be used with `autoUpdate`', () => {
    const { Wrapper, chain } = createChain();
    const ref: { current: UseEmojiReturn | null } = { current: null };

    const Component = () => {
      useRemirrorContext({ autoUpdate: true });
      ref.current = useEmoji();

      return null;
    };

    strictRender(
      <Wrapper>
        <Component />
      </Wrapper>,
    );

    renderAct(() => {
      chain.insertText(':');
    });

    expect(ref.current).toEqual(expect.objectContaining({ state: expect.any(Object) }));
  });
});
