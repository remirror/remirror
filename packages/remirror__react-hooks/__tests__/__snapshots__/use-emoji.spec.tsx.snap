// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useEmoji should choose selection when \`Enter\` key is pressed 1`] = `
<p>
  Initial content
  <span
    class="remirror-emoji-wrapper"
    data-remirror-emoji="😂"
  >
    <img
      role="presentation"
      class="remirror-emoji-image"
      aria-label="face with tears of joy"
      alt="face with tears of joy"
      src="https://cdn.jsdelivr.net/npm/@svgmoji/noto@3.2.0/sprites/all.svg#1F602"
    >
  </span>
  <img
    class="ProseMirror-separator"
    alt
  >
  <br class="ProseMirror-trailingBreak">
</p>
`;

exports[`useEmoji should correctly add the emoji when the command is called 1`] = `
<p>
  Initial content
  <span
    class="remirror-emoji-wrapper"
    data-remirror-emoji="❤️"
  >
    <img
      role="presentation"
      class="remirror-emoji-image"
      aria-label="red heart"
      alt="red heart"
      src="https://cdn.jsdelivr.net/npm/@svgmoji/noto@3.2.0/sprites/all.svg#2764"
    >
  </span>
  <img
    class="ProseMirror-separator"
    alt
  >
  <br class="ProseMirror-trailingBreak">
</p>
`;
